narwhal/primary/src/state_handler.rs

```diff
@@ -270,14 +280,14 @@ impl RandomnessState {
                         error!("random beacon: unable to write VSS key to output: {e:?}")
                     }
                     let num_shares = output.shares.as_ref().map_or(0, |shares| shares.len());
-                    self.dkg_output = Some(output);
+                    self.store.set_dkg_output(output);
```

Comment:
How do you figure that missed messages won't make it to state handler? After starting back up post-crash, don't missed certificates all have to be sent through bullshark, or else Sui will lose a bunch of consensus output?  and the code that sends sequenced certs to the state handler comes right after the code that sends it to the executor. it seemed to me like this should work, can you help illustrate the flow where stuff gets missed?

Reply:
There is a consensus recovery process, but this includes:
* the bullshark DAG state, taking into account unprocessed certificates from the mempool (see [here](https://github.com/MystenLabs/sui/blob/3af3e0a353ff3ce699a6be9b0ddbf8c05b61892d/narwhal/primary/src/consensus/state.rs#L65))
* the missed messages by the Executor based on the last executed message (this is marked by SUI) as you already mentioned

and the above suffice to ensure that the current bullshark state and SUI are able to continue from where left off. 

However, no attempt is made to restore any of the consumer of the [tx_committed_certificates](https://github.com/MystenLabs/sui/blob/3af3e0a353ff3ce699a6be9b0ddbf8c05b61892d/narwhal/primary/src/consensus/state.rs#L453) - which is the state handler. We have no information of the latest height that state handler has processed messages up to and it was never crucial for the recovery process either.

So to make the recovery process work as expected I believe in the simplest form we should keep track of the latest processed committed message by the state handler (just the sub dag index would be enough) and try to re-process all the messages up to the latest commit index.

Reply:
As far as I can tell from the code, when narwhal restarts, it will load the last committed subdag from the store, and then wait until it gets the next new certificate from consensus, send it through bullshark as normal. that will cause bullshark to trigger commit of any previous subdags not yet committed, which will include sending that to the state handler via tx_committed_certificates.

Are you saying there is some other code path where on startup it will send committed subdags to Sui without going through the `Consensus::run_inner` loop? If so can you please point me to it? (the recovery code you linked just loads saved state but does not trigger any commits AFAICT)